core/src/main/scala/kafka/log/LogSegment.scala

```diff
@@ -125,26 +125,77 @@ class LogSegment private[log] (val log: FileRecords,
              largestTimestamp: Long,
              shallowOffsetOfMaxTimestamp: Long,
              records: MemoryRecords): Unit = {
+    append(firstOffset, largestOffset, largestTimestamp, shallowOffsetOfMaxTimestamp, records,
+           false)
+  }
+
+  /**
+   * Append the given messages starting with the given offset. Add
+   * an entry to the index if needed.
+   *
+   * It is assumed this method is being called from within a lock.
+   *
+   * @param firstOffset The first offset in the message set.
+   * @param largestOffset The last offset in the message set
+   * @param largestTimestamp The largest timestamp in the message set.
+   * @param shallowOffsetOfMaxTimestamp The offset of the message that has the largest timestamp in the messages to append.
+   * @param records The log entries to append.
+   * @param mayBeLegacySegment The segment might be a pre KAFKA-5413 "legacy" segment, which
```

Comment:
I think I still prefer to use a name which is more suggestive of what it actually achieves (e.g. `allowOversizeOffset`). An alternative would be to let this option be implicit in the name of the method? For example, we could have `appendToLogOnly` which skips the index append. That is probably sufficient to avoid accidental misuse and seems like a cleaner API.

Reply:
I feel `append()` and `appendToLogOnly()` is somewhat confusing. I understand the intent but if I did not know the code and was trying to figure out which one to use, I would first need to understand what "log only" and "index" means. Given that we only have two callers for the append() function, this might not be too bad, but it still reveals concepts like "overflow" and "index" that should ideally be abstracted under the log segment.

I prefer the current naming because there is no possibility of misuse - we don't want clients other than log cleaner to be bothered about what `allowOversizeOffset` might mean, and neither do we want clients to pass in "true" for that parameter by mistake.
`mayBeLegacySegment` is a pretty good name in that sense I feel, and most clients, just by looking at the parameter name, would know what to set it to.

In any case, I think it's okay to have a log-private function with `allowOversizeOffset` while still retaining the public append() API that currently exists. That should hopefully keep things cleaner.

Reply:
I think my gripe is primarily that the notion of a "legacy segment" is difficult to explain. It assumes a context in which we are copying from an existing segment, which is only true for the log cleaner. I would have no idea what it means unless I read the big blurb beside the parameter name, and even then, I need to consult the JIRA to understand the context. Basically we are letting this workaround leak into the Log API. We can't help that to some extent to fix this issue, but at least we can try to keep the abstraction clean. Having a way to append to the log without appending to the index is easy to understand without knowing the details of this bug, and it could even be useful in testing.